pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -178,21 +181,64 @@ func (c *EtcdCertSignerController) syncAllMasters(recorder events.Recorder) erro
 	return utilerrors.NewAggregate(errs)
 }
 
+// validateCertificateSAN makes sure that a given certificate is valid
+// at least for the SANs defined in the configuration.
+func verifyCertificateSAN(certData map[string][]byte, certName string, ipAddresses []string) error {
+	certificates, err := cert.ParseCertsPEM(certData["tls.crt"])
+	if err != nil {
+		return fmt.Errorf("could not parse TLS cert for %s: %w", certName, err)
+	}
+	// Using the first certificate to verify the IP addresses
+	certificate := certificates[0]
+	for _, ipAddress := range ipAddresses {
+		if err := certificate.VerifyHostname(ipAddress); err != nil {
+			return fmt.Errorf("SAN for the certificate %s does not include %s: %w", certName, ipAddress, err)
+		}
+	}
+	return nil
+}
+
 func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node, recorder events.Recorder) error {
 	etcdPeerClientCertName := tlshelpers.GetPeerClientSecretNameForNode(node.Name)
 	etcdServingCertName := tlshelpers.GetServingSecretNameForNode(node.Name)
 	metricsServingCertName := tlshelpers.GetServingMetricsSecretNameForNode(node.Name)
 
-	var err error
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
-	peerClientCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdServingCertName)
-	servingCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(metricsServingCertName)
-	metricsCertOk := err == nil
+	certNames := []string{etcdPeerClientCertName, etcdServingCertName, metricsServingCertName}
+
+	nodeInternalIPs, err := dnshelpers.GetInternalIPAddressesForNodeName(node)
+	if err != nil {
+		return err
+	}
+
+	errors := []error{}
+	secrets := []*corev1.Secret{}
+	for _, certName := range certNames {
+		secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
+		if err != nil {
+			errors = append(errors, err)
+			break
+		} else {
+			if secret.Data == nil {
+				errors = append(errors, fmt.Errorf("secret data not found in %s", certName))
+			}
+			secrets = append(secrets, secret)
+		}
+	}
 
-	// if we have all the certs we want, do nothing.
-	if peerClientCertOk && servingCertOk && metricsCertOk {
+	// If we found all the secrets, match the IP SAN.
+	if len(errors) == 0 {
+		sanMismatchErrs := []error{}
+		for index, certName := range certNames {
```

Comment:
why do we need two loops?

Reply:
Because the first loop determines if all the secrets are available. If any of them is not available, then, you immediately jump to regenerating certs.

If all certs are available, then, validate the SAN. If there is a SAN mismatch returning degrading error (no regeneration).

Reply:
I originally had a single loop for checking the existence of the certs and then validating the SAN. 

The logic for determining to generate certs, I felt, became complicated and less readable. This way, it appears cleaner and more maintainable.